File: Euclid/Network/Codec/NetworkDecoder.cs
using DotNetty.Buffers;
using DotNetty.Codecs;
using DotNetty.Transport.Channels;
using Euclid.Network.Streams;
using Euclid.Util;
using System.Collections.Generic;

namespace Euclid.Network
{
    internal class NetworkDecoder : ByteToMessageDecoder
    {
        protected override void Decode(IChannelHandlerContext ctx, IByteBuffer buffer, List<object> output)
        {
            if (buffer.ReadableBytes < 5)
            {
                // If the incoming data is less than 6 bytes, it's junk.
                return;
            }

            byte[] message_length = buffer.ReadBytes(4).Array;
            int.TryParse(StringUtil.GetEncoding().GetString(message_length).Trim(), out int length);

            byte[] message = buffer.ReadBytes(length).Array;
            string content = StringUtil.GetEncoding().GetString(message);

            string request = string.Empty;
            string header;

            if (content.Contains(" "))
            {
                header = content.Split(" ")[0];
                request = content.Substring(header.Length + 1);
            }
            else
            {
                header = content;
            }

            output.Add(new Request(header, request));
        }
    }
}

File: euclid/src/main/java/euclid/network/codec/NetworkDecoder.java
package euclid.network.codec;

import euclid.network.streams.Request;
import euclid.util.StringUtil;
import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.ByteToMessageDecoder;

import java.util.List;

public class NetworkDecoder extends ByteToMessageDecoder {

    @Override
    protected void decode(ChannelHandlerContext ctx, ByteBuf buffer, List<Object> output) {
        if (buffer.readableBytes() < 5) {
            return;
        }

        byte[] lengthBytes = new byte[4];
        buffer.readBytes(lengthBytes);
        int length = Integer.parseInt(new String(lengthBytes, StringUtil.getEncoding()).trim());

        byte[] message = new byte[length];
        buffer.readBytes(message);
        String content = new String(message, StringUtil.getEncoding());

        String header;
        String request;

        if (content.contains(" ")) {
            header = content.split(" ", 2)[0];
            request = content.substring(header.length() + 1);
        } else {
            header = content;
            request = "";
        }

        output.add(new Request(header, request));
    }
}
